### PR TITLE
[GEOS-9512]: RasterDownload on Heterogeneous CRS ImageMosaics

### PR DESCRIPTION
**Pull request by dromagnoli**
<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Committs changing the REST API, or any configuration object, should check it the REST API docs (Swagger YAML files and classic documentation) need to be updated.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

File: src/community/wps-download/src/test/java/org/geoserver/wps/gs/download/DownloadProcessTest.java [hunk 1]
```diff
@@ -1379,6 +1398,102 @@ public void testDownloadScaledRasterPreservingDatatype() throws Exception {
         }
     }
 
+    /**
+     * Test download of raster data. The source is an ImageMosaic with Heterogeneous CRS. Sending a
+     * request with a TargetCRS matching one of the underlying CRS of that mosaic should result in
+     * minimal reprocessing on granules with that CRS as native.
+     */
+    @Test
+    public void testDownloadGranuleHeterogeneousCRS() throws Exception {
```
Comment (aaime) [inline src/community/wps-download/src/test/java/org/geoserver/wps/gs/download/DownloadProcessTest.java]: Tests only the happy path. What about a case where the bbox is not catching the target CRS? Or having a target width and height set?

File: doc/en/user/source/community/wps-download/rawDownload.rst [hunk 2]
```diff
@@ -425,4 +426,38 @@ The supported writing parameters are:
...
+
+By default, mosaicking granules with different CRSs will result into a reprojection from the original CRS of the granules to that common CRS.
+
+Since 2.17, a new parameter has been defined: ``bestRawOnMatchingCrs``
+
+It will be used on Raster Download with a defined ROI and a TargetCRS without target size being specified. When set to true, the download will aim to preserve the native properties of the underlying granules matching the targetCRS, as much as possible.
+
+Back to the example, a RasterDownload specifying UTM32N as TargetCRS and a ROI covering an area containing UTM32N Granules will result into getting those UTM32N granules without applying any intermediate reprojection, also providing back best raw resolution available for that CRS. So, if the ImageMosaic is a mix of UTM32N DEMs at 10 km resolution and UTM33N at 100 m resolution, the underlying reading request will use 10 km resolution, being the best resolution available in the targetCRS.
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply doc/en/user/source/community/wps-download/rawDownload.rst]: yep. if multiple granules are in the ROI and both of them are in target CRS and they have different resolution, than the best one get used.

File: src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/RasterDownload.java [hunk 3]
```diff
@@ -223,9 +238,9 @@ public Resource execute(
             ROIManager roiManager = null;
             if (roi != null) {
                 roiManager = new ROIManager(roi, (CoordinateReferenceSystem) roi.getUserData());
-
+                roiManager.useBestRawForTargetCRS = useBestRawForTargetCRS;
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/RasterDownload.java]: Will fix it

File: src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/GridGeometryProvider.java [hunk 4]
```diff
@@ -103,22 +109,29 @@ public ReferencedEnvelope getBestResolution(
             CoordinateReferenceSystem schemaCrs =
                     schema.getGeometryDescriptor().getCoordinateReferenceSystem();
 
+            CoordinateReferenceSystem referenceCrs =
+                    useBestRawForTargetCRS ? roiManager.getTargetCRS() : schemaCrs;
             // Iterate over the features to extract the best resolution
             BoundingBox featureBBox = null;
             GeneralEnvelope env = null;
             ReferencedEnvelope envelope = null;
+            double[] fallbackResolution =
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/GridGeometryProvider.java]: There is an explanation about fallback vs best on the validateresolution method. Should I move that here?
This private method won't be invoked at all if the feature collection is empty.

File: src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/DownloadProcess.java [hunk 5]
```diff
@@ -174,6 +173,14 @@ public File execute(
                         min = 0
                     )
                     Parameters writeParameters,
+            @DescribeParameter(
+                        name = "bestRawOnMatchingCrs",
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/DownloadProcess.java]: > This name needs to be modified. So you pick the best resolution from the matching target CRS? If so, maybe "bestResolutionOnMatchingCRS" (note the capitalization, matches the other parameters, targetCRS and roiCRS).
> And this activates only if targetSizeX and targetSizeY are empty right? Whatever the logic, it should be documented. Wondering if an exception should be thrown too?

I have put the detailed documentation in the GeoServer doc page above. Would that be enough?
Maybe developers won't check that page and only look at the javadoc, so probably better to duplicate it here too.

> 
> But... even if the target size is set, one might still like not having the double reprojection, or not?
> This parameter really is mixing two things:
> 
> * avoid double reprojecting the granules that are already in the target CRS
> * using the resolution of the granules in the target CRS as the guidance for the native resolution
>   Maybe it should be split in two?

The documentation was indeed referring about the typical use case of need to get something near to raw download of the "original" data. So specifying a ROI containing some granules and their CRS and get them back basically unmodified (so using same resolution without reprojection).

I was thinking that if you accept to get them at a different resolution you aren't "targeting" to the original "imageLayout". Would that make sense?

**Comment by dromagnoli** [comment]
All feedbacks have been addressed with:
https://github.com/geoserver/geoserver/pull/4113. 
Closing this one

**Pull request closed without merging**